src/plugins/data/common/utils/abort_utils.ts

```diff
@@ -36,29 +36,47 @@ export class AbortError extends Error {
  *
  * @param signal The `AbortSignal` to generate the `Promise` from
  */
-export function toPromise(signal: AbortSignal): Promise<never> {
-  return new Promise((resolve, reject) => {
-    if (signal.aborted) reject(new AbortError());
-    const abortHandler = () => {
+export function toPromise(signal: AbortSignal): Promise<never> & { cleanup: () => void } {
```

Comment:
Hmm, I'm not a huge fan of adding a method to a built-in class (`Promise`). We might consider returning an object instead (like `{ promise, cleanup }`). I think it will also help to enforce that any consumers of this function will handle cleaning up and not just use the promise returned.

Reply:
I don't mind initial approach I took mostly because it is clearly typed that it has a `cleanup` method, nevertheless: 

>  I think it will also help to enforce that any consumers of this function will handle cleaning up and not just use the promise returned.

I agree with this 👍 changed `{promise, cleanup}` 

one tiny caveat is that now we can't just use:

```
Promise.race(signals.map(toPromise))
```

but have to:

```
Promise.race(signals.map(toPromise).map(pr => pr.promise))
```